fedimint-client/src/lib.rs

```diff
@@ -1356,6 +1376,201 @@ impl Client {
+    ///
+    /// This will block until the recovery task is done with recoveries.
+    /// Returns success if all recovery tasks are complete (success case),
+    /// or an error if some modules could not complete the recovery at the time.
+    ///
+    /// A bit of a heavy approach.
+    pub async fn wait_for_all_recoveries(&self) -> anyhow::Result<()> {
+        let mut recovery_receiver = self.client_recovery_progress_receiver.clone();
+        recovery_receiver
+            .wait_for(|in_progress| {
+                !in_progress
+                    .iter()
+                    .any(|(_id, progress)| !progress.is_done())
+            })
+            .await
+            .context("Recovery task completed, but some modules could not recover")?;
+
+        Ok(())
+    }
+
+    pub async fn wait_for_module_kind_recovery(
+        &self,
+        module_kind: ModuleKind,
+    ) -> anyhow::Result<()> {
+        let mut recovery_receiver = self.client_recovery_progress_receiver.clone();
+        recovery_receiver
+            .wait_for(|in_progress| {
+                !in_progress
+                    .iter()
+                    .filter(|(module_instance_id, _progress)| {
+                        self.config.modules[module_instance_id].kind == module_kind
+                    })
+                    .any(|(_id, progress)| !progress.is_done())
+            })
+            .await
+            .context("Recovery task completed, but some modules could not recover")?;
+
+        Ok(())
+    }
+
+    pub async fn wait_for_all_active_state_machines(&self) -> anyhow::Result<()> {
+        loop {
+            if self.executor.get_active_states().await.is_empty() {
+                break;
+            }
+            fedimint_core::task::sleep(Duration::from_millis(100)).await;
+        }
+        Ok(())
+    }
+
+    /// Set the client [`Metadata`]
+    pub async fn set_metadata_dbtx(dbtx: &mut DatabaseTransaction<'_>, metadata: &Metadata) {
+        dbtx.insert_new_entry(&ClientMetadataKey, metadata).await;
+    }
+
+    async fn spawn_module_recoveries_task(
+        &self,
+        recovery_sender: watch::Sender<BTreeMap<ModuleInstanceId, RecoveryProgress>>,
+        module_recoveries: BTreeMap<
+            ModuleInstanceId,
+            Pin<Box<maybe_add_send!(dyn Future<Output = anyhow::Result<()>>)>>,
+        >,
+        module_recovery_progress_receivers: BTreeMap<
+            ModuleInstanceId,
+            watch::Receiver<RecoveryProgress>,
+        >,
+    ) {
+        let db = self.db.clone();
+        self.task_group
+            .spawn("module recoveries", move |_task_handle| async move {
+                Self::run_module_recoveries_task(
+                    db,
+                    recovery_sender,
+                    module_recoveries,
+                    module_recovery_progress_receivers,
+                )
+                .await
+            })
+            .await;
+    }
+
+    async fn run_module_recoveries_task(
```

Comment:
shouldn't there be a client.modules.register_module call after module's recovery is complete?
or we are meant to restart client after recovery is complete.

Reply:
The client is already `Arc`ed, potentially working etc. It would introduce a lot of locking and complications that might be a source of mistakes (hard to reason about things when module set changes underneath), so I think the limitation that recovered modules start being usable only after next client start is a good tradeoff.